### PR TITLE
Implement edx problem interaction events pydantic models

### PR DESCRIPTION
**Pull request by quitterie-lcs**
# Purpose

Problem interaction events are `Edx` events generated during platform usage. They are emitted by the server or the browser to capture information about interactions with problems.

# Proposal

Write `pydantic` models, edge-case and selector tests for:

- [x] edx.problem.hint.demandhint_displayed
- [x] edx.problem.hint.feedback_displayed
- [x] problem_check (browser)
- [x] problem_check (server)
- [x] problem_check_fail
- [x] problem_graded
- [x] problem_rescore
- [x] problem_rescore_fail
- [x] reset_problem
- [x] reset_problem_fail
- [x] save_problem_fail
- [x] save_problem_success
- [x] showanswer

### CODE REVIEW

File: src/ralph/models/edx/problem_interaction/fields/events.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+    Attributes:
+        answers (json): Consists of a dictionary of problem ID and the internal answer identifier
+            for each problem.
+        failure (str): Consists of either the `closed` or `unreset` value.
+        problem_id (str): Consists of the ID of the problem that was checked.
+        state (json): Consists of the current problem state.
+    """
+
+    # pylint: disable=unsubscriptable-object
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/edx/problem_interaction/fields/events.py:206]: It is necessary when using `Json` from `pydantic`

[the remark this answers is not in the document]
Comment (jmaupetit) [reply src/ralph/models/edx/problem_interaction/fields/events.py:206]: Can you add a comment? :pray:

**Review by jmaupetit — COMMENTED** [review]
LGTM. Waiting for @SergioSim 's approval on this.

File: src/ralph/models/edx/problem_interaction/fields/events.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/edx/problem_interaction/fields/events.py:206]: We already did that in the other models. Do we have to mention it on every `statements` file?

[the remark this answers is not in the document]
Comment (jmaupetit) [reply src/ralph/models/edx/problem_interaction/fields/events.py:206]: I have mixed feelings about this...

**Review by SergioSim — COMMENTED** [review]
This is awesome! Great work! I have only minor suggestions and questions.
image

**Comment by SergioSim** [comment]
Apropos, I pasted a couple of problem interaction events [here](https://pastebin.com/iAkdQuJe). We could use them for validation:
`cat /path/to/events.txt | ./bin/ralph -v DEBUG validate -f edx -F`

File: src/ralph/models/edx/problem_interaction/statements.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    SaveProblemFailEventField,
+    SaveProblemSuccessEventField,
+    ShowAnswerEventField,
+    UIProblemCheckEventField,
+    UIProblemResetEventField,
+    UIProblemShowEventField,
+)
+
+
+class EdxProblemHintDemandhintDisplayed(BaseServerModel):
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/edx/problem_interaction/statements.py:30]: Thanks for this crucial information. I did not pay attention to it. 
I would prefer to use an optional module field rather than creating an additional model.

File: src/ralph/models/edx/problem_interaction/fields/events.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+    module_id: str
+    problem_part_id: str
+    question_type: Union[
+        Literal["stringresponse"],
+        Literal["choiceresponse"],
+        Literal["multiplechoiceresponse"],
+        Literal["numericalresponse"],
+        Literal["optionresponse"],
+    ]
+    student_answer: Union[str, List[str]]
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/edx/problem_interaction/fields/events.py:137]: I rechecked, it is always a list even with one string.

File: src/ralph/models/edx/problem_interaction/statements.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+    Attributes:
+        event (json): See ProblemGradedEventField.
+        event_type (str): Consists of the value `problem_graded`.
+    """
+
+    __selector__ = selector(event_source="server", event_type="problem_graded")
+
+    event: List[ProblemGradedEventField]
+    event_type: Literal["problem_graded"]
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/edx/problem_interaction/statements.py:136]: Yes it is. Good catch of this mistake :hand:

File: src/ralph/models/edx/problem_interaction/statements.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+class UIProblemSave(BaseBrowserModel):
+    """Represents the `problem_save` browser event.
+
+    Attributes:
+        event_type (str): Consists of the value `problem_save`.
+    """
+
+    __selector__ = selector(event_source="browser", event_type="problem_save")
+
+    event_type: Literal["problem_save"]
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/edx/problem_interaction/statements.py:199]: Is it necessary to add `event` field ? 
https://edx.readthedocs.io/projects/devdata/en/latest/internal_data_formats/tracking_logs/student_event_types.html#problem-save

[the remark this answers is not in the document]
Comment (SergioSim) [reply src/ralph/models/edx/problem_interaction/statements.py:199]: The `event` field seems to contain an url-encoded string. (see [display.coffee](https://github.com/openfun/edx-platform/blob/master/common/lib/xmodule/xmodule/js/src/capa/display.coffee#L415))

File: src/ralph/models/edx/problem_interaction/fields/events.py [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+    Attributes:
+        correctness (str): Consists either of the `correct` or `incorrect` value.
+        hint (str): Consists of optional hint.
+        hint_mode (str): Consists either of the value `on_request` or `always` value.
+        msg (str): Consists of extra message response.
+        npoints (int): Consists of awarded points.
+        queuestate (json): see QueueStateField.
+    """
+
+    correctness: Union[Literal["correct"], Literal["incorrect"]]
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply src/ralph/models/edx/problem_interaction/fields/events.py:35]: Where did you find it?

[the remark this answers is not in the document]
Comment (SergioSim) [reply src/ralph/models/edx/problem_interaction/fields/events.py:35]: I think it's in [correctmap.py](https://github.com/openfun/edx-platform/blob/master/common/lib/capa/capa/correctmap.py#L54-L61)

File: src/ralph/models/edx/base.py [hunk 7]
```diff
@@ -15,6 +15,22 @@ class Config:  # pylint: disable=missing-class-docstring
...
+        usage_key (str): Consists of a block ID of the current component.
+        display_name (str): Consists of a short description or title of the component.
+    """
+
+    usage_key: constr(regex=r"^block-v1:.+\+.+\+.+type@.+@[a-f0-9]{32}$")  # noqa:F722
+    display_name: str
+    original_usage_key: constr(
+        regex=r"^block-v1:.+\+.+\+.+type@problem\+block@[a-f0-9]{32}$"  # noqa:F722
+    )
+    original_usage_version: str
```
Comment (SergioSim) [inline src/ralph/models/edx/base.py:31]:
```suggestion
    original_usage_key: Optional[constr(
        regex=r"^block-v1:.+\+.+\+.+type@problem\+block@[a-f0-9]{32}$"  # noqa:F722
    )]
    original_usage_version: Optional[str]
```
The `original_usage_key` and `original_usage_version` fields are not always present

File: src/ralph/models/edx/problem_interaction/statements.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+
+    Attributes:
+        event (list): See ProblemGradedEventField.
+        event_type (str): Consists of the value `problem_graded`.
+        name (str): Consists of the value `problem_graded`.
+    """
+
+    __selector__ = selector(event_source="browser", event_type="problem_graded")
+
+    event: List[UIProblemGradedEventField]
```
Comment (SergioSim) [inline src/ralph/models/edx/problem_interaction/statements.py:135]:
```suggestion
    event: List[Union[str, Literal[None], None]]
```
The event field holds a list of strings (generally of length 2 but not 100% sure)
I tried to document it [here](https://github.com/SergioSim/ralph/blob/docs/docs/browserLogs.md#event_type-problem_graded).

**Review by SergioSim — COMMENTED** [review]
Almost there! Only 3 minor propositions)
Then all the test events [here](https://pastebin.com/iAkdQuJe) should pass without errors.
`cat /path/to/events.txt | ./bin/ralph -v DEBUG validate -f edx -F`
image

**Review by SergioSim — APPROVED** [review]
This looks great!
image
